### PR TITLE
support char mode emoji and multibyte emoji like 👨‍👩‍👧‍👦

### PR DESCRIPTION
**Pull request by qiuguohua**
https://github.com/cocos/cocos-engine/issues/15192
https://github.com/cocos/cocos-engine/pull/16687#pullrequestreview-1887694982

Re: #

### Changelog

1. Optimize character expression processing
2. Add unit test

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: tests/ui/test-utils.test.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    //  图形 = utf-16 编码 = 10进制 = 连接字符串
+    //  🐻 = 0xDB3D 0xDC38 = 55357 56379 = 5535756379;
+    //  👨‍👩‍👧‍👦 = 0xDB3D 0xDC68 0x200D 0xDB3D 0xDC69 0x200D 0xDB3D 0xDC67 0x200D 0xDB3D 0xDC66 = 55357 56424 8205 55357 56425 8205 55357 56423 8205 55357 56422 = 5535756424820555357564258205553575642382055535756422;
+    //  🐼 = 0xDB3D 0xDC3C = 55357 56380 = 5535756380
+    expect(getSymbolCodeAt(txt, 3)).toStrictEqual('5535756379'); // 🐻
+    expect(getSymbolCodeAt(txt, 4)).toStrictEqual('5535756424820555357564258205553575642382055535756422'); // 👨‍👩‍👧‍👦
+    expect(getSymbolCodeAt(txt, 5)).toStrictEqual('5535756380'); // 🐼
+});
+
+// test('getSymbolStartIndex', function () {
```
Comment (minggo) [inline tests/ui/test-utils.test.ts]: Why comment these test cases?

File: tests/ui/test-utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    expect(getSymbolAt(txt, 5)).toStrictEqual('🐼');
+    
+});
+
+test('getSymbolCodeAt', function () {
+    let txt = 'w大熊🐻👨‍👩‍👧‍👦🐼';
+    expect(getSymbolCodeAt(txt, 0)).toStrictEqual('119');
+    expect(getSymbolCodeAt(txt, 1)).toStrictEqual('22823');
+    expect(getSymbolCodeAt(txt, 2)).toStrictEqual('29066');
+    //  图形 = utf-16 编码 = 10进制 = 连接字符串
```
Comment (minggo) [inline tests/ui/test-utils.test.ts]: Please don't use Chinese.

File: tests/ui/test-utils.test.ts (hunk 1 above)
Comment (qiuguohua) [reply tests/ui/test-utils.test.ts]: These functions are not exported, I exported them temporarily for testing

File: tests/ui/test-utils.test.ts (hunk 2 above)
Comment (qiuguohua) [reply tests/ui/test-utils.test.ts]: ok

File: tests/ui/test-utils.test.ts (hunk 1 above)
Comment (minggo) [reply tests/ui/test-utils.test.ts]: Got it, could you please add some comments.

Comment (qiuguohua) [reply tests/ui/test-utils.test.ts]: ok

**Comment by minggo** [comment]
Please fix eslint error as possible.

**Review by minggo — APPROVED** [review]

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by qiuguohua** [comment]
https://github.com/cocos/cocos-engine/issues/16710

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Pull request merged**